File: core/src/com/desitum/library/game/World.java
package com.desitum.library.game;

import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.desitum.library.game_objects.GameObject;
import com.desitum.library.particles.ParticleEmitter;
import com.desitum.library.view.TouchEvent;
import com.desitum.library.view.View;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by kody on 12/27/15.
 * can be used by kody and people in [kody}]
 */
public class World implements InputProcessor{

    public boolean mLayerGameObjects = true;
    private boolean mClickDown = false;
    private boolean mForegroundClickDown = false;

    private List<GameObject> mGameObjects;
    private List<G2DSprite> mSprites;
    private List<View> mViews;
    private List<ParticleEmitter> mParticleEmitters;
    private OrthographicCamera mCamera;
    private OrthographicCamera mForegroundCamera;
    private Viewport mViewport;
    private Viewport mForegroundViewport;
    private View mViewFocus;
    private TouchEvent mTouchEvent;

    /**
     * Create new {@link World}
     * @param camera camera from {@link GameScreen}
     */
    public World(OrthographicCamera camera, Viewport viewport) {
        this.mCamera = camera;
        this.mViewport = viewport;
        this.mGameObjects = new ArrayList<GameObject>();
        this.mSprites = new ArrayList<G2DSprite>();
        this.mViews = new ArrayList<View>();
        this.mParticleEmitters = new ArrayList<ParticleEmitter>();
        this.mTouchEvent = new TouchEvent();
    }

    /**
     * Create new {@link World}
     * @param camera camera from {@link GameScreen}
     */
    public World(OrthographicCamera camera, Viewport viewport, OrthographicCamera foregroundCamera, Viewport foregroundViewport) {
        this.mCamera = camera;
        this.mViewport = viewport;
        this.mForegroundCamera = foregroundCamera;
        this.mForegroundViewport = foregroundViewport;
        this.mGameObjects = new ArrayList<GameObject>();
        this.mSprites = new ArrayList<G2DSprite>();
        this.mViews = new ArrayList<View>();
        this.mParticleEmitters = new ArrayList<ParticleEmitter>();
        this.mTouchEvent = new TouchEvent();
    }

    /**
     * update the world every frame using this method
     * @param delta time since last update
     */
    public void update(float delta) {
        for (GameObject gameObject : mGameObjects) {
            gameObject.update(delta);
        }
        for (ParticleEmitter particleEmitter : mParticleEmitters) {
            particleEmitter.update(delta);
        }
        for (G2DSprite g2DSprite : mSprites) {
            g2DSprite.update(delta);
        }
        for (G2DSprite g2DSprite : mViews) {
            g2DSprite.update(delta);
        }
    }

    /**
     * updates the world based on touch input
     * for best results update before the master update
     * @param touchPos position of mouse or touch point if on mobile
     * @param touchDown is clicking or if currently touching screen
     */
    public boolean updateTouchInput(Vector3 touchPos, boolean touchDown) {
        boolean returnVal = false;
        mTouchEvent.setX(touchPos.x);
        mTouchEvent.setY(touchPos.y);
        if (touchDown && !mClickDown) {
            returnVal = onTouchDown(touchPos);
        } else if (!touchDown && mClickDown) {
            returnVal = onTouchUp();
        } else if (touchDown && mClickDown) {
            returnVal = onTouchMoved();
        }
        for (GameObject gameObject2D : mGameObjects) {
            gameObject2D.updateTouchInput(touchPos, touchDown);
        }
        mClickDown = touchDown;
        return returnVal;
    }

    /**
     * updates the world based on touch input
     * for best results update before the master update
     * @param touchPos position of mouse or touch point if on mobile
     * @param touchDown is clicking or if currently touching screen
     */
    public boolean updateForegroundTouchInput(Vector3 touchPos, boolean touchDown) {
        boolean returnValue = false;
        if (!mViews.isEmpty()) {
            mTouchEvent.setX(touchPos.x);
            mTouchEvent.setY(touchPos.y);
            if (touchDown && !mForegroundClickDown) {
                returnValue = onTouchDownForeground();
            } else if (!touchDown && mForegroundClickDown) {
                returnValue = onTouchUpForeground();
            } else if (touchDown && mForegroundClickDown) {
                returnValue = onTouchMovedForeground();
            }
            mForegroundClickDown = touchDown;
        }
        return returnValue;
    }

    /**
     * Add a {@link GameObject} to be handled by {@link World}
     * @param gameObject2D {@link GameObject} to be handled
     */
    public void addGameObject(GameObject gameObject2D) {
        this.mGameObjects.add(gameObject2D);
        if (mLayerGameObjects) {
            Collections.sort(mGameObjects);
        }
    }

    /**
     * Add a {@link ParticleEmitter} to be handled by {@link World}
     * @param particleEmitter {@link ParticleEmitter} to be handled
     */
    public void addParticleEmitter(ParticleEmitter particleEmitter) {
        this.mParticleEmitters.add(particleEmitter);
    }

    /**
     * Add a {@link View} to be handled by {@link World}
     * @param view {@link View} to be handled
     */
    public void addView(View view) {
        this.mViews.add(view);
        Collections.sort(mViews);
    }

    /**
     * Remove a game object from the screen
     * @param gameObject {@link GameObject} to remove
     */
    public void removeGameObject(GameObject gameObject) {
        mGameObjects.remove(gameObject);
    }

    /**
     * Remove a view from the screen
     * @param view {@link View} to remove
     */
    public void removeView(View view) {
        mViews.remove(view);
    }

    /**
     * Get a {@link List} of {@link GameObject}
     * @return {@link List} of {@link GameObject}
     */
    public List<GameObject> getGameObjects() {
        return mGameObjects;
    }

    /**
     * Get a {@link List} of {@link View}
     * @return {@link List} of {@link View}
     */
    public List<View> getViews() {
        return mViews;
    }

    /**
     * Get a {@link List} of {@link G2DSprite}
     * @return {@link List} of {@link G2DSprite}
     */
    public List<G2DSprite> getSprites() {
        return mSprites;
    }

    /**
     * Get {@link OrthographicCamera} used by the {@link World}
     * @return {@link OrthographicCamera}
     */
    public OrthographicCamera getCamera() {
        return mCamera;
    }

    /**
     * Get {@link OrthographicCamera} used by the {@link World}
     * @return {@link OrthographicCamera}
     */
    public OrthographicCamera getForegroundCamera() {
        return mForegroundCamera;
    }

    public boolean onTouchDown(Vector3 clickPos) {
        mTouchEvent.setAction(TouchEvent.Action.DOWN);
        // TODO Game Object Touch Events
        return false;
    }

    public boolean onTouchUp() {
        mTouchEvent.setAction(TouchEvent.Action.UP);
        // TODO Game Object Touch Events
        return false;
    }

    private boolean onTouchMoved() {
        mTouchEvent.setAction(TouchEvent.Action.MOVE);
        // TODO Game Object Touch Events
        return false;
    }

    public boolean onTouchDownForeground() {
        mTouchEvent.setAction(TouchEvent.Action.DOWN);
        for (View view : mViews) {
            if (view.isTouching(mTouchEvent)) {
                view.dispatchTouchEvent(mTouchEvent);
                return true;
            }
        }
        return false;
    }

    public boolean onTouchUpForeground() {
        mTouchEvent.setAction(TouchEvent.Action.UP);
        if (mViewFocus != null) {
            mViewFocus.dispatchTouchEvent(mTouchEvent);
            mViewFocus = null;
            return true;
        }
        return false;
    }

    private boolean onTouchMovedForeground() {
        mTouchEvent.setAction(TouchEvent.Action.MOVE);
        if (mViewFocus != null) {
            mViewFocus.dispatchTouchEvent(mTouchEvent);
            return true;
        }
        return false;
    }

    public void setCamera(OrthographicCamera camera) {
        this.mCamera = camera;
    }

    public Viewport getViewport() {
        return mViewport;
    }

    public void setViewport(Viewport viewport) {
        this.mViewport = viewport;
    }

    public Viewport getForegroundViewport() {
        return mForegroundViewport;
    }

    public void dispose() {
        for (ParticleEmitter particleEmitter: mParticleEmitters) {
            particleEmitter.dispose();
        }
        for (GameObject gameObject: mGameObjects) {
            gameObject.dispose();
        }
    }

    public List<ParticleEmitter> getParticleEmitters() {
        return mParticleEmitters;
    }

    public void requestFocus(View view) {
        for (View v : mViews) {
            v.clearFocus();
        }
        mViewFocus = view;
        mViewFocus.setFocus(true);
    }

    @Override
    public boolean keyDown(int keycode) {
        return false;
    }

    @Override
    public boolean keyUp(int keycode) {
        return false;
    }

    @Override
    public boolean keyTyped(char character) {
        return false;
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button) {
        return false;
    }

    @Override
    public boolean touchUp(int screenX, int screenY, int pointer, int button) {
        return false;
    }

    @Override
    public boolean touchDragged(int screenX, int screenY, int pointer) {
        return false;
    }

    @Override
    public boolean mouseMoved(int screenX, int screenY) {
        return false;
    }

    @Override
    public boolean scrolled(int amount) {
        return false;
    }
}

File: core/src/com/desitum/library/game/World.kt
package com.desitum.library.game

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.viewport.Viewport
import com.desitum.library.game_objects.GameObject
import com.desitum.library.particles.ParticleEmitter
import com.desitum.library.view.TouchEvent
import com.desitum.library.view.View

import java.util.ArrayList
import java.util.Collections

/**
 * Created by kody on 12/27/15.
 * can be used by kody and people in [kody}]
 */
class World
/**
 * Create new [World]
 * @param camera camera from [GameScreen]
 */(var camera: OrthographicCamera, var viewport: Viewport, var foregroundCamera: OrthographicCamera, var foregroundViewport: Viewport) {

    var layerGameObjects = true
    private var clickDown = false
    private var foregroundClickDown = false

    var gameObjects: MutableList<GameObject>
        private set
    var sprites: List<G2DSprite>
        private set
    var views: MutableList<View>
        private set
    var particleEmitters: MutableList<ParticleEmitter>
        private set

    private var viewFocus: View? = null
    private var gameObjectFocus: GameObject? = null
    private var touchEvent: TouchEvent

    /**
     * update the world every frame using this method
     * @param delta time since last update
     */
    fun update(delta: Float) {
        gameObjects.forEach { it.update(delta) }
        particleEmitters.forEach { it.update(delta) }
        sprites.forEach { it.update(delta) }
        views.forEach { it.update(delta) }
    }

    /**
     * updates the world based on touch input
     * for best results update before the master update
     * @param touchPos position of mouse or touch point if on mobile
     * *
     * @param touchDown is clicking or if currently touching screen
     */
    fun updateTouchInput(touchPos: Vector3, touchDown: Boolean): Boolean {
        var returnVal = false
        touchEvent.x = touchPos.x
        touchEvent.y = touchPos.y
        if (touchDown && !clickDown) {
            returnVal = onTouchDown()
        } else if (!touchDown && clickDown) {
            returnVal = onTouchUp()
        } else if (touchDown && clickDown) {
            returnVal = onTouchMoved()
        }
//        gameObjects.forEach { it.updateTouchInput(touchPos, touchDown) }
        clickDown = touchDown
        return returnVal
    }

    /**
     * updates the world based on touch input
     * for best results update before the master update
     * @param touchPos position of mouse or touch point if on mobile
     * *
     * @param touchDown is clicking or if currently touching screen
     */
    fun updateForegroundTouchInput(touchPos: Vector3, touchDown: Boolean): Boolean {
        var returnValue = false
        if (!views.isEmpty()) {
            touchEvent.x = touchPos.x
            touchEvent.y = touchPos.y
            if (touchDown && !foregroundClickDown) {
                returnValue = onTouchDownForeground()
            } else if (!touchDown && foregroundClickDown) {
                returnValue = onTouchUpForeground()
            } else if (touchDown && foregroundClickDown) {
                returnValue = onTouchMovedForeground()
            }
            foregroundClickDown = touchDown
        }
        return returnValue
    }

    /**
     * Add a [GameObject] to be handled by [World]
     * @param gameObject2D [GameObject] to be handled
     */
    fun addGameObject(gameObject2D: GameObject) {
        this.gameObjects.add(gameObject2D)
        if (layerGameObjects) {
            Collections.sort(gameObjects)
        }
    }

    /**
     * Add a [ParticleEmitter] to be handled by [World]
     * @param particleEmitter [ParticleEmitter] to be handled
     */
    fun addParticleEmitter(particleEmitter: ParticleEmitter) {
        this.particleEmitters.add(particleEmitter)
    }

    /**
     * Add a [View] to be handled by [World]
     * @param view [View] to be handled
     */
    fun addView(view: View) {
        this.views.add(view)
        Collections.sort(views)
    }

    /**
     * Remove a game object from the screen
     * @param gameObject [GameObject] to remove
     */
    fun removeGameObject(gameObject: GameObject) {
        gameObjects.remove(gameObject)
    }

    /**
     * Remove a view from the screen
     * @param view [View] to remove
     */
    fun removeView(view: View) {
        views.remove(view)
    }

    fun onTouchDown(): Boolean {
        touchEvent.action = TouchEvent.Action.DOWN
        gameObjects.reversed()
                .filter { it.isTouching(touchEvent) }
                .forEach { it.dispatchTouchEvent(touchEvent) }
        return false
    }

    fun onTouchUp(): Boolean {
        touchEvent.action = TouchEvent.Action.UP
        if (gameObjectFocus != null) {
            gameObjectFocus!!.dispatchTouchEvent(touchEvent)
            gameObjectFocus = null
            return true
        }
        return false
    }

    private fun onTouchMoved(): Boolean {
        touchEvent.action = TouchEvent.Action.MOVE
        return gameObjectFocus?.dispatchTouchEvent(touchEvent) ?: false
    }

    fun onTouchDownForeground(): Boolean {
        touchEvent.action = TouchEvent.Action.DOWN
        return views.reversed()
                .filter { it.isTouching(touchEvent) }
                .firstOrNull()?.dispatchTouchEvent(touchEvent) ?: false
    }

    fun onTouchUpForeground(): Boolean {
        touchEvent.action = TouchEvent.Action.UP
        if (viewFocus != null) {
            viewFocus!!.dispatchTouchEvent(touchEvent)
            viewFocus = null
            return true
        }
        return false
    }

    private fun onTouchMovedForeground(): Boolean {
        touchEvent.action = TouchEvent.Action.MOVE
        return viewFocus?.dispatchTouchEvent(touchEvent) ?: false
    }

    fun dispose() {
        for (particleEmitter in particleEmitters) {
            particleEmitter.dispose()
        }
        for (gameObject in gameObjects) {
            gameObject.dispose()
        }
    }

    fun requestFocus(view: View) {
        for (v in views) {
            v.clearFocus()
        }
        viewFocus = view
        viewFocus!!.focus = true
    }

    fun requestFocus(gameObject: GameObject) {
        gameObjects.forEach { it.clearFocus() }
        gameObjectFocus = gameObject
        gameObjectFocus!!.focus = true
    }

    init {
        this.gameObjects = ArrayList<GameObject>()
        this.sprites = ArrayList<G2DSprite>()
        this.views = ArrayList<View>()
        this.particleEmitters = ArrayList<ParticleEmitter>()
        this.touchEvent = TouchEvent()
    }
}
